storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java

```diff
@@ -1025,21 +1062,48 @@ private void removeNonCanonicalBlobSidecars(
       final Map<Bytes32, UInt64> deletedHotBlocks,
       final Map<Bytes32, Bytes32> finalizedChildToParentMap) {
     try (final FinalizedUpdater updater = finalizedUpdater()) {
-      LOG.trace("Removing blob sidecars for non-canonical blocks");
       final Set<SlotAndBlockRoot> nonCanonicalBlocks =
           deletedHotBlocks.entrySet().stream()
               .filter(entry -> !finalizedChildToParentMap.containsKey(entry.getKey()))
               .map(entry -> new SlotAndBlockRoot(entry.getValue(), entry.getKey()))
               .collect(Collectors.toSet());
-      for (final SlotAndBlockRoot slotAndBlockRoot : nonCanonicalBlocks) {
-        dao.getBlobSidecarKeys(slotAndBlockRoot)
-            .forEach(
-                key -> {
-                  LOG.trace("Removing blobSidecar with index {} for non-canonical block", key);
-                  updater.removeBlobSidecar(key);
-                });
+      if (storeNonCanonicalBlocks) {
```

Comment:
I think this is broken for 2 reasons:
- You cannot commit multiple times an updater.
- `index == nonCanonicalBlocks.size() - 1` not holds if you are actually on the second round. Maybe we can use `!nonCanonicalBlocksIterator.hasNext()` condition to check if we are at last element.


try to put `BLOBS_TX_BATCH_SIZE = 2` and run the tests.

Reply:
Fixed it and picked your test improvments. Thanks!